### PR TITLE
add SetChainID() to caller interface

### PR DESCRIPTION
**Pull request by dustinxie**
PR consists 2 commits:
1. simplify call interface -- removed unnecessary interfaces, merged common methods to `SendActionCaller` interface
2. add SetChainID() interface

### CODE REVIEW

File: iotex/caller_claimreward.go [hunk 1]
```diff
@@ -11,56 +11,29 @@ import (
...
-	return c
-}
-
-func (c *claimRewardCaller) SetNonce(n uint64) ClaimRewardCaller {
-	c.nonce = n
-	return c
-}
-
-func (c *claimRewardCaller) API() iotexapi.APIServiceClient { return c.api }
-
```
Comment (dustinxie) [inline iotex/caller_claimreward.go:51]: common codes, merged to `sendActionCaller`

File: iotex/caller_contract.go [hunk 2]
```diff
@@ -35,23 +35,6 @@ func (c *deployContractCaller) SetArgs(abi abi.ABI, args ...interface{}) DeployC
...
-	return c
-}
-
-func (c *deployContractCaller) SetNonce(n uint64) DeployContractCaller {
-	c.nonce = n
-	return c
-}
-
-func (c *deployContractCaller) API() iotexapi.APIServiceClient { return c.api }
-
```
Comment (dustinxie) [inline iotex/caller_contract.go:54]: common codes

File: iotex/caller_contract.go [hunk 3]
```diff
@@ -98,23 +81,6 @@ func (c *executeContractCaller) SetAmount(a *big.Int) ExecuteContractCaller {
...
-	return c
-}
-
-func (c *executeContractCaller) SetNonce(n uint64) ExecuteContractCaller {
-	c.nonce = n
-	return c
-}
-
-func (c *executeContractCaller) API() iotexapi.APIServiceClient { return c.api }
-
```
Comment (dustinxie) [inline iotex/caller_contract.go:117]: common codes

File: iotex/caller_staking.go [hunk 4]
```diff
@@ -33,131 +33,103 @@ type (
...
-	return c
-}
-
-//SetPayload set basic data
-func (c *stakingCaller) SetPayload(pl []byte) StakingAPICaller {
-	if pl == nil {
-		return c
-	}
-	c.payload = make([]byte, len(pl))
-	copy(c.payload, pl)
```
Comment (dustinxie) [inline iotex/caller_staking.go:160]: common codes

File: iotex/caller_transfer.go [hunk 5]
```diff
@@ -12,58 +12,31 @@ import (
...
-	return c
-}
-
-func (c *transferCaller) SetNonce(n uint64) TransferCaller {
-	c.nonce = n
-	return c
-}
-
-func (c *transferCaller) API() iotexapi.APIServiceClient { return c.api }
-
```
Comment (dustinxie) [inline iotex/caller_transfer.go:53]: common codes

File: iotex/interfaces.go [hunk 6]
```diff
@@ -12,30 +12,27 @@ import (
...
-	SetData([]byte) ClaimRewardCaller
-	SetNonce(uint64) ClaimRewardCaller
+// SendActionCaller is used to set nonce/gas and derive other callers.
+type SendActionCaller interface {
+	Caller
+	SetNonce(uint64) SendActionCaller
+	SetGasLimit(uint64) SendActionCaller
+	SetGasPrice(*big.Int) SendActionCaller
+	SetChainID(uint32) SendActionCaller
+	SetPayload([]byte) SendActionCaller
```
Comment (dustinxie) [inline iotex/interfaces.go:28]: move the common `SetXXX` method to `SendActionCaller`

File: iotex/interfaces.go [hunk 7]
```diff
@@ -12,30 +12,27 @@ import (
...
-}
-
-// ClaimRewardCaller is used to perform a claim reward call.
-type ClaimRewardCaller interface {
-	SendActionCaller
-
-	SetGasPrice(*big.Int) ClaimRewardCaller
-	SetGasLimit(uint64) ClaimRewardCaller
-	SetData([]byte) ClaimRewardCaller
-	SetNonce(uint64) ClaimRewardCaller
```
Comment (dustinxie) [inline iotex/interfaces.go:38]: these 2 are no longer needed

**Comment by Liuhaai** [comment]
The heavy refactoring work should be separated into another pr

**Comment by dustinxie** [comment]
> The heavy refactoring work should be separated into another pr

OK, created PR103, once 103 merge, will rebase this one

**Review by Liuhaai — APPROVED** [review]

**Pull request merged**